### PR TITLE
provides racing/corner-cases verification and test coverage

### PR DESCRIPTION
**Pull request by OlegDokuka**

### CODE REVIEW

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java [hunk 1]
```diff
@@ -276,14 +266,18 @@ void drainSync() {
                 if (v == null) {
                     try {
                         a.onCompleted();
-                    } catch (Throwable throwable) {
-                        cancel();
-                        a.onError(prepareError(throwable));
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Why was this guard deleted?

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java [hunk 2]
```diff
@@ -389,16 +389,11 @@ void drainRegular() {
...
-                    } catch (Throwable ignore) {
-                        cancel();
-                    }
+                    } catch (Throwable ignore) { }
                 } else {
                     try {
                         a.onCompleted();
-                    } catch (Throwable throwable) {
-                        cancel();
-                        a.onError(prepareError(throwable));
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Why was this guard deleted?

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java [hunk 3]
```diff
@@ -389,16 +389,11 @@ void drainRegular() {
                 if (t != null) {
                     try {
                         a.onError(prepareError(t));
-                    } catch (Throwable ignore) {
-                        cancel();
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Why was this guard deleted?

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java [hunk 4]
```diff
@@ -433,17 +428,12 @@ boolean checkTerminated(boolean d, boolean empty, CallStreamObserver<?> a, Queue
...
-                    cancel();
-                }
+                } catch (Throwable ignore) { }
                 return true;
             } else if (empty) {
                 try {
                     a.onCompleted();
-                } catch (Throwable throwable) {
-                    cancel();
-                    a.onError(prepareError(throwable));
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Why was this guard deleted?

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java [hunk 5]
```diff
@@ -433,17 +428,12 @@ boolean checkTerminated(boolean d, boolean empty, CallStreamObserver<?> a, Queue
                 q.clear();
                 try {
                     a.onError(prepareError(t));
-                } catch (Throwable ignore) {
-                    cancel();
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Why was this guard deleted?

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java [hunk 6]
```diff
@@ -459,4 +449,115 @@ private static Throwable prepareError(Throwable throwable) {
         }
     }
 
+    /**
+     * Implementation of Cancelled Queue Subscription which is used as a marker of
+     * cancelled {@link AbstractSubscriberAndProducer} instance.
+     */
+    private static class CancelledQueueSubscription implements Subscription, Queue {
+
+        static final String NOT_SUPPORTED_MESSAGE = "Although CancelledQueueSubscription implements Queue it is" +
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Message not needed since class is `private static`. There is no way for a user to instantiate this class.

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java [hunk 7]
```diff
@@ -152,23 +144,23 @@ public void onSubscribe(Subscription subscription) {
...
+        if (t == null && sourceMode == ASYNC || sourceMode == NOT_FUSED) {
             drain();
             return;
         }
 
         if (!isCanceled()) {
             checkNotNull(t);
 
-            final CallStreamObserver<T> stream = downstream;
+            final CallStreamObserver<T> a = downstream;
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Please find a more descriptive variable name than `a`?

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package com.salesforce.reactivegrpc.common;
+
+import io.reactivex.FlowableSubscriber;
+import io.reactivex.internal.fuseable.QueueSubscription;
+import org.reactivestreams.Subscription;
+
+public class TestSubscriberProducer<T> extends AbstractSubscriberAndProducer<T>
```
Comment (rmichela) [inline common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java]: Please document what this class does.

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package com.salesforce.reactivegrpc.common;
```
Comment (rmichela) [inline common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java]: Needs license header

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducerTest.java [hunk 10]
```diff
@@ -0,0 +1,583 @@
+package com.salesforce.reactivegrpc.common;
```
Comment (rmichela) [inline common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducerTest.java]: Needs license header

**Comment by rmichela** [comment]
Does this address #139?

**Comment by OlegDokuka** [comment]
This work is in progress. Will ammend later today

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java (hunk 1 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: Due to java docs, this any further interactions with Observer are forbidden -> https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/StreamObserver.java#L75

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java (hunk 2 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: the same -> https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/StreamObserver.java#L75

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java (hunk 7 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: done

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java (hunk 6 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducer.java]: fixed

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java (hunk 9 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java]: added

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java (hunk 8 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/TestSubscriberProducer.java]: added docs

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducerTest.java (hunk 10 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducerTest.java]: added

**Comment by rmichela** [comment]
Both CI runs hung on `com.salesforce.reactivegrpc.common.StreamObserverAndPublisherTest`

**Comment by OlegDokuka** [comment]
Fixing the number of rounds. It looks like my local machine have a little bit higher performance characteristics, so those tests do not cause any problems from the time-lasting perspective :D
